File sets tutorial (#802)
NixOS/nix.dev: source/tutorials/file-sets.md

To build a local project in a Nix derivation, its source files must be accessible to the builder.
But since the builder runs in an isolated environment (if the [sandbox](https://nixos.org/manual/nix/stable/command-ref/conf-file.html#conf-sandbox) is enabled),
it won't have access to the local project files by default.

To make this work regardless, the Nix language has certain builtin features to copy local paths to the Nix store,
whose paths are then accessible to derivation builders [^1].

[^1]: Technically only Nix store paths from the derivations inputs can be accessed,
but in practice this distinction is not important.

Using these features directly can be tricky however:

- Coercion of paths to strings, such as the wide-spread pattern of `src = ./.`,
  makes the derivation dependent on the name of the current directory.
  It also doesn't allow being more precise about which files to use.

- The [`builtins.path`](https://nixos.org/manual/nix/stable/language/builtins.html#builtins-path) function
  (and equivalently [`lib.sources.cleanSourceWith`](https://nixos.org/manual/nixpkgs/stable/#function-library-lib.sources.cleanSourceWith))

>   can address these problems.
>   However, it's hard to get the desired path selection using function-based interface.

Author: Good point! I'd word it a bit more lightly though. While it is technically old, it's not deprecated. How about

```suggestion
  can address these problems.
  However, it's hard to get the desired path selection using the `filter` function interface.
```

>   However, it's hard to get the desired path selection using function-based interface.

Reviewer: Seems good to me!

(file-sets)=
# Working with local files
<!-- TODO: Switch all mentions of unstable to stable once 23.11 is out
Also remove mention of using unstable
Also use https://nixos.org/manual/nixpkgs/unstable/#sec-functions-library-fileset
instead of https://nixos.org/manual/nixpkgs/unstable/#sec-fileset
-->

> To build a local project in a Nix derivation,
> its source files must be accessible
> to the [derivation builder](https://nixos.org/manual/nix/stable/language/derivations#attr-builder).
> Since by default, derivations get built in an [isolated environment](https://nixos.org/manual/nix/stable/command-ref/conf-file.html#conf-sandbox)
> that only allows reading from the Nix store,
> the Nix language has built-in features to copy local files to the store and expose the resulting store paths.

Reviewer 2: ```suggestion
To build a local project in a Nix derivation, source files must be accessible to its [`builder` executable](https://nixos.org/manual/nix/stable/language/derivations#attr-builder).
Since by default, the `builder` runs in an [isolated environment](https://nixos.org/manual/nix/stable/command-ref/conf-file.html#conf-sandbox) that only allows reading from the Nix store, the Nix language has built-in features to copy local files to the store and expose the resulting store paths.
```

I think it's not wrong to keep these simple mechanisms transparent. I won't fight though, and leave it as an optional suggestion.

Such path arguments are then [implicitly turned into sets](https://nixos.org/manual/nixpkgs/unstable/#sec-fileset-path-coercion) that contain _all_ files under the given path.
In the previous trace this is indicated by `(all files in directory)`.

:::{tip}
The `trace` function pretty-prints its first argument and returns its second argument.
But since you often just need the pretty-printing in `nix repl`, you can omit the second argument:

```shell-session
nix-repl> fs.trace ./.
trace: /home/user (all files in directory)
«lambda @ /nix/store/1czr278x24s3bl6qdnifpvm5z03wfi2p-nixpkgs-src/lib/fileset/default.nix:555:8»
```
:::

Even though file sets conceptually contain local files, these files are *never* added to the Nix store unless explicitly requested.
So You don't have to worry as much about accidentally copying secrets into the world-readable store.

In this example, although we pretty-printed the home directory, no files were copied.
This is in contrast to coercion of paths to strings such as in `"${./.}"`,
which copies the whole directory to the Nix store on evaluation!

:::{warning}

> When using [experimental Flakes](https://nixos.org/manual/nix/stable/command-ref/new-cli/nix3-flake), local directories with a `flake.nix` are always copied into the Nix store *completely* unless it is a Git repository!

Reviewer 2: What are "experimental Flakes"? Are there non-experimental flakes?

I still think that should be "With the `flakes` experimental feature enabled, ..."

Author: > When using the [`flakes` and `nix-command` experimental features](https://nixos.org/manual/nix/stable/command-ref/new-cli/nix3-flake),
a local directory within a Flake is always copied into the Nix store *completely* unless it is a Git repository!

In the meeting

:::

This implicit coercion also works for files:

```shell-session
$ touch some-file
```

```shell-session
nix-repl> fs.trace ./some-file
trace: /home/user
trace: - some-file (regular)
```

In addition to the included file, this also prints its [file type](https://nixos.org/manual/nix/stable/language/builtins.html#builtins-readFileType).

## Example project

To further experiment with the library, make a sample project.
Create a new directory, enter it,
and set up `niv` to pin the Nixpkgs dependency:

```shell-session
$ mkdir fileset
$ cd fileset
$ nix-shell -p niv --run "niv init --nixpkgs nixos/nixpkgs --nixpkgs-branch master"
```

Then create a `default.nix` file with the following contents:

```{code-block} nix
:caption: default.nix
{
  system ? builtins.currentSystem,
  sources ? import ./nix/sources.nix,
}:
let
  pkgs = import sources.nixpkgs {
    config = { };
    overlays = [ ];
    inherit system;
  };
in
pkgs.callPackage ./build.nix { }
```

Add two source files to work with:

```shell-session
$ echo hello > hello.txt
$ echo world > world.txt
```

## Adding files to the Nix store

> To add files in a given file set to the Nix store,
> we use [`toSource`](https://nixos.org/manual/nixpkgs/unstable/#function-library-lib.fileset.toSource).

Reviewer 2: Who is "we" here? It makes sense to say "we" when we're referring to this group writing the tutorial, e.g. when explaining why we're telling people to fetch from `master` rather than `nixpkgs-unstable`. But here here we'd run the danger of saying "we use this, you do you", while we really want to either state a fact "X does Y" or tell people what to do.
```suggestion
Files in a given file set can be added to the Nix store with [`toSource`](https://nixos.org/manual/nixpkgs/unstable/#function-library-lib.fileset.toSource).
```

I don't super like the phrasing, but it should be good enough.

> This function needs a `root` attribute to know which path to use as the root of the resulting store path,
> but only exactly the files in the `fileset` argument are included in the result.

Reviewer 2: "root of store path" is potentially confusing. Isn't that `/nix/store`?

There is no "but" here,  because one part doesn't contradict the other.

```suggestion
The argument to this function requires a `root` attribute to determine which source directory to copy to the store, and exactly the files in the `fileset` attribute are included in the result.
```

PS: I really don't like breaking sentences up into multiple lines, because edits are usually based on sentence, not phrase. Don't know about others, but dealing with multiple lines in the GitHub UI almost always make it more tedious for me.

Author: Used
> The argument to this function requires a `root` attribute to determine which source directory to copy to the store.
Only the files in the `fileset` attribute are included in the result.

Define `build.nix` as follows:

```{code-block} nix
:caption: build.nix
{ stdenv, lib }:
let
  fs = lib.fileset;
  sourceFiles = ./hello.txt;
in

fs.trace sourceFiles

stdenv.mkDerivation {
  name = "fileset";
  src = fs.toSource {
    root = ./.;
    fileset = sourceFiles;
  };
  postInstall = ''
    mkdir $out
    cp -v hello.txt $out
  '';
}
```

The call to `fs.trace` prints the file set that will be used as a derivation input.

> The build will succeed (note that it will take some time to fetch the first time around):

Reviewer 2: This should take some time already when doing the first `trace`, so either note it there or skip.

```suggestion
Try building it:
```

Author: Happens here due to niv not being able to re-use the download.

Changed it to an admonition though (see commit)

trace: - string.txt (regular)
this derivation will be built:
  /nix/store/zmgpqlpfz2jq0w9rdacsnpx8ni4n77cn-filesets.drv
...
/nix/store/6pffjljjy3c7kla60nljk3fad4q4kkzn-filesets
```

One way to fix this is to use [`unions`](https://nixos.org/manual/nixpkgs/unstable/#function-library-lib.fileset.unions).

Create a file set containing a union of the files to exclude (`fs.unions [ ... ]`), and subtract it (`difference`) from the complete directory (`./.`):

```{code-block} diff
:caption: build.nix
 { stdenv, lib }:
 let
   fs = lib.fileset;
-  sourceFiles = fs.difference ./. (fs.maybeMissing ./result);
+  sourceFiles =
+    fs.difference
+      ./.
+      (fs.unions [
+        (fs.maybeMissing ./result)
+        ./default.nix
+        ./build.nix
+        ./nix
+      ]);
 in
```

Changing any of the excluded files now doesn't necessarily cause a new build anymore:

```
$ echo >> build.nix
```

```
$ nix-build
trace: /home/user/fileset
trace: - hello.txt (regular)
trace: - world.txt (regular)
/nix/store/ckn40y7hgqphhbhyrq64h9r6rvdh973r-fileset
```

## Filter

With the [`fileFilter`](https://nixos.org/manual/nixpkgs/unstable/#function-library-lib.fileset.fileFilter) function,

> you can find sets of files satisfying a criteria.

Author: > The [`fileFilter`](https://nixos.org/manual/nixpkgs/unstable/#function-library-lib.fileset.fileFilter) function allows filtering file sets such that each included file satisfies the given criteria.

 { stdenv, lib }:
 let
   fs = lib.fileset;
   sourceFiles =
     fs.difference
       ./.
       (fs.unions [
         (fs.maybeMissing ./result)
-        ./default.nix
-        ./build.nix
+        (fs.fileFilter (file: lib.hasSuffix ".nix" file.name) ./.)
         ./nix
       ]);
 in
```

This does not change the result, even if we add a new `.nix` file.

```shell-session
$ nix-build
trace: /home/user/fileset
trace: - hello.txt (regular)
trace: - world.txt (regular)
/nix/store/ckn40y7hgqphhbhyrq64h9r6rvdh973r-fileset
```

Notably, the approach of using `difference ./.` explicitly selects the files to _exclude_, which means that new files added to the source directory are included by default.
Depending on your project, this might be a better fit than the alternative in the next section.

## Union (explicitly include files)

To contrast the previous approach, `unions` can also be used to select only the files to _include_.
This means that new files added to the current directory would be ignored by default.

Create some additional files:

```shell-session
$ mkdir src
$ touch build.sh src/select.{c,h}
```

Then create a file set from only the files to be included explicitly:

> ```{code-block} diff

Author: Changed it back to a partial file, no diff.

     fileset = sourceFiles;
   };
   postInstall = ''
-    mkdir $out
-    cp -v {hello,world}.txt $out
+    cp -vr . $out
   '';
 }
```

The `postInstall` script is simplified to rely on the sources to be pre-filtered appropriately:

```shell-session
$ nix-build
trace: /home/user/fileset
trace: - build.sh (regular)
trace: - hello.txt (regular)
trace: - src (all files in directory)
trace: - world.txt (regular)
this derivation will be built:
  /nix/store/sjzkn07d6a4qfp60p6dc64pzvmmdafff-fileset.drv
...
'.' -> '/nix/store/zl4n1g6is4cmsqf02dci5b2h5zd0ia4r-fileset'
'./build.sh' -> '/nix/store/zl4n1g6is4cmsqf02dci5b2h5zd0ia4r-fileset/build.sh'
'./hello.txt' -> '/nix/store/zl4n1g6is4cmsqf02dci5b2h5zd0ia4r-fileset/hello.txt'
'./world.txt' -> '/nix/store/zl4n1g6is4cmsqf02dci5b2h5zd0ia4r-fileset/world.txt'
'./src' -> '/nix/store/zl4n1g6is4cmsqf02dci5b2h5zd0ia4r-fileset/src'
'./src/select.c' -> '/nix/store/zl4n1g6is4cmsqf02dci5b2h5zd0ia4r-fileset/src/select.c'
'./src/select.h' -> '/nix/store/zl4n1g6is4cmsqf02dci5b2h5zd0ia4r-fileset/src/select.h'
...
/nix/store/zl4n1g6is4cmsqf02dci5b2h5zd0ia4r-fileset
```

Only the specified files are used, even when a new one is added:

```shell-session

> $ touch src/select.o README.md

Author: discussed in meeting, it's fine to have it

-      )
-      ./src
-    )
-  ];
+  sourceFiles = fs.gitTracked ./.;
 in
```

Building we get

```shell-session
$ nix-build
warning: Git tree '/home/user/fileset' is dirty
trace: /home/vg/src/nix.dev/fileset
trace: - README.md (regular)
trace: - build.nix (regular)
trace: - build.sh (regular)
trace: - default.nix (regular)
trace: - hello.txt (regular)
trace: - nix (all files in directory)
trace: - src
trace:   - select.c (regular)
trace:   - select.h (regular)
trace: - world.txt (regular)
this derivation will be built:
  /nix/store/p9aw3fl5xcjbgg9yagykywvskzgrmk5y-fileset.drv
...
/nix/store/cw4bza1r27iimzrdbfl4yn5xr36d6k5l-fileset
```

This includes too much though, as not all of these files are needed to build the derivation as originally intended.

:::{note}
When using [experimental Flakes](https://nixos.org/manual/nix/stable/command-ref/new-cli/nix3-flake), it's [not really possible](https://github.com/NixOS/nix/issues/9292) to use this function, even with

```shell-session
$ nix build path:.
```

However it's also not needed, because by default, `nix build` only allows access to files tracked by Git.
:::

## Intersection

This is where `intersection` comes in.

> It allows creating a file set that consists only of files that are in _both_ of two given file sets.

Author: Discussed, it's fine without :)

```shell-session
$ nix-build
trace: /home/user/fileset
trace: - build.nix (regular)
trace: - default.nix (regular)
trace: - hello.txt (regular)
trace: - nix (all files in directory)
trace: - world.txt (regular)
this derivation will be built:
  /nix/store/zr19bv51085zz005yk7pw4s9sglmafvn-fileset.drv
...
'hello.txt' -> '/nix/store/vhyhk6ij39gjapqavz1j1x3zbiy3qc1a-fileset/hello.txt'
'world.txt' -> '/nix/store/vhyhk6ij39gjapqavz1j1x3zbiy3qc1a-fileset/world.txt'
...
/nix/store/vhyhk6ij39gjapqavz1j1x3zbiy3qc1a-fileset
```

An attempt to repeat the build will re-use the existing store path:

```
$ nix-build
trace: /home/user/fileset
trace: - build.nix (regular)
trace: - default.nix (regular)
trace: - hello.txt (regular)
trace: - nix (all files in directory)
trace: - world.txt (regular)
/nix/store/vhyhk6ij39gjapqavz1j1x3zbiy3qc1a-fileset
```

## Missing files

Removing the `./result` symlink creates a new problem, though:

```shell-session
$ rm result
$ nix-build
error: lib.fileset.difference: Second argument (negative set)
  (/home/user/select/result) is a path that does not exist.
  To create a file set from a path that may not exist, use `lib.fileset.maybeMissing`.
```

> Follow the instructions in the error message, and use `maybeMissing` to create a file set from a path that may not exist (in which case the file set will be empty):

Author: Note that this is not yet merged: https://github.com/NixOS/nixpkgs/pull/265964

So to make this work for now, you'll have to use that branch instead:
```
nix-shell -p niv --run "niv update nixpkgs --owner tweag --branch fileset.optional"
```

We shouldn't merge this PR until at least that PR is merged.

Reviewer 3: Unblocked

(file-sets)=
# Working with local files
<!-- TODO: Switch all mentions of unstable to stable once 23.11 is out -->

> To build a local project in a Nix derivation, its source files must be accessible to the [`builder` executable](https://nixos.org/manual/nix/stable/language/derivations#attr-builder).

Reviewer 2: I disagree. I would have to think hard what "hermetic" is even supposed to mean. An executable being able to read files is something very concrete and down-to-earth in Unix land.

> To build a local project in a Nix derivation, its source files must be accessible to the [`builder` executable](https://nixos.org/manual/nix/stable/language/derivations#attr-builder).

Author: Hmm I agree with @fricklerhandwerk here, there's exactly one mention of "hermetic" in the Nix manual, and it doesn't tell you what it means.

Though I also agree with @roberth that it's a bit too deep, `builder` is irrevelant to pretty much everybody using Nix.

How about this:
```suggestion
To build a local project in a Nix derivation,
its source files must be [accessible](https://nixos.org/manual/nix/stable/command-ref/conf-file.html#conf-sandbox)
to the [derivation builder](https://nixos.org/manual/nix/stable/language/derivations#attr-builder).
```

> To build a local project in a Nix derivation, its source files must be accessible to the [`builder` executable](https://nixos.org/manual/nix/stable/language/derivations#attr-builder).

Reviewer 2: I insist that "`builder` executable" is unambiguous as opposed to everything else. "derivation builder" is jargon that may as well refer to `mkDerivation` (the rename to "build helper" is very recent) or a remote machine.

Such path arguments are then [implicitly coerced](https://nixos.org/manual/nixpkgs/unstable/#sec-fileset-path-coercion), and the resulting file sets contain _all_ files under the given path.
In the previous trace this is indicated by `(all files in directory)`.

:::{tip}
The `trace` function pretty-prints its first agument and returns its second argument.
But since you often just need the pretty-printing in `nix repl`, you can omit the second argument:

```shell-session
nix-repl> fs.trace ./.
trace: /home/user (all files in directory)
«lambda @ /nix/store/1czr278x24s3bl6qdnifpvm5z03wfi2p-nixpkgs-src/lib/fileset/default.nix:555:8»
```
:::

Even though file sets conceptually contain local files, these files are *never* added to the Nix store unless explicitly requested.
You don't have to worry about accidentally copying secrets into the world-readable store.

In this example, although we pretty-printed the home directory, no files were copied.
This is also evident from the expression evaluating instantly.

:::{note}
This is in contrast to coercion of paths to strings such as in `"${./.}"`,
which copies the whole directory to the Nix store on evaluation!

> :::

Reviewer 2: Agreed, let's stay focus and leave room for an "all the ways to handle files" reference.

> :::

Author: I don't think it hurts to include this though, it's just one sentence but it still teaches something related. I now combined it with the above paragraph though, that sounds good.

:::{warning}
With [experimental Flakes](https://nixos.org/manual/nix/stable/command-ref/new-cli/nix3-flake), a local directory containing `flake.nix` is always copied into the Nix store *completely* unless it is a Git repository!
:::

This implicit coercion also works for files:

```shell-session
nix-repl> fs.trace /etc/nix/nix.conf
trace: /etc/nix
trace: - nix.conf (symlink)

> ```

Reviewer 2: I suspect the motivation here was, before having any setup, how to show something the library does? nix.conf won't be a symlink on a vanilla Ubuntu install, but that's a minor issue.

> ```

Author: Indeed, `/etc/nix/nix.conf` was the only file I could think of that every user with Nix should have

> ```

Author: I now changed this to just creating a file beforehand: [(diff)](https://github.com/NixOS/nix.dev/pull/802/commits/67b688adb6addc00a0a7af482add9460ea6acae5)

In addition to the included file, this also prints its [file type](https://nixos.org/manual/nix/stable/language/builtins.html#builtins-readFileType).

If a given path doesn't exist, the library will complain:

```shell-session
nix-repl> fs.trace /etc/nix/nix.nix
error: lib.fileset.trace: Argument (/etc/nix/nix.nix)
  is a path that does not exist.

> ```

Reviewer 2: It's fair to make users aware of this, since there's a good chance they will introduce typos. It's not wrong to have it but I won't complain if it was removed.

## Example project

To further experiment with the library, make a sample project.
Create a new directory, enter it,
and set up `niv` to pin the Nixpkgs dependency:

```shell-session
$ mkdir fileset
$ cd fileset
$ nix-shell -p niv --run "niv init --nixpkgs nixos/nixpkgs --nixpkgs-branch nixos-unstable"
```

:::{note}

> We're using the `nixos-unstable` channel branch here, since no stable release has all the features needed for this tutorial.

Reviewer 3: ```suggestion
We're using the `nixos-unstable` channel branch here, because the library is part of NixOS 23.11 and up, which has not been released as of writing.
```

Reviewer 2: ```suggestion
We're using the `nixos-unstable` channel branch here, because the library is part of NixOS 23.11 and up, which has not been released at the of writing.
```

:::

Then create a `default.nix` file with the following contents:

```{code-block} nix
:caption: default.nix
{
  system ? builtins.currentSystem,

>   inputs ? import ./nix/sources.nix,
> }:
> let
>   pkgs = import inputs.nixpkgs {

Reviewer 2: ```suggestion
  sources ? import ./nix/sources.nix,
}:
let
  pkgs = import sources.nixpkgs {
```

As discussed in #806 this was a fluke on my end.

do not know how to unpack source archive /nix/store/i9pmrzmpshapij2kin22pff6fc2adavx-
hello.txt
error: builder for '/nix/store/kaik07l2i7s7wh7lcxpwynv4lmn4h742-fileset.drv' failed w
ith exit code 1;
       last 3 log lines:
       > unpacking sources
       > unpacking source archive /nix/store/i9pmrzmpshapij2kin22pff6fc2adavx-hello.t
xt
       > do not know how to unpack source archive /nix/store/i9pmrzmpshapij2kin22pff6
fc2adavx-hello.txt
       For full logs, run 'nix log /nix/store/kaik07l2i7s7wh7lcxpwynv4lmn4h742-filese
t.drv'.
```

## Adding files to the Nix store

That's where [`toSource`](https://nixos.org/manual/nixpkgs/unstable/#function-library-lib.fileset.toSource) comes in:
It adds exactly the files in the given file set to a directory in the Nix store, starting from a specified root path.

Define `build.nix` as follows:

```{code-block} nix
:caption: build.nix
{ stdenv, lib }:
let
  fs = lib.fileset;
  sourceFiles = ./hello.txt;
in

fs.trace sourceFiles

stdenv.mkDerivation {
  name = "fileset";
  src = fs.toSource {
    root = ./.;
    fileset = sourceFiles;
  };
  postInstall = ''
    mkdir $out
    cp -v hello.txt $out
  '';
}
```

> The call to `fs.trace` will print the file set that will be used as a derivation input.

Reviewer 3: ```suggestion
The call to `fs.trace` prints the file set that will be used as a derivation input. We'll remove it later.
```

Reviewer 2: We won't in this tutorial?

Author: Yeah, I'll change the `print` though.

  To create a file set from a path that may not exist, use `lib.fileset.maybeMissing`.
```

Follow the instructions in the error message, and use `maybeMissing` to create a file set from a path that may not exist (in which case the file set will be empty):

<!-- https://nixos.org/manual/nixpkgs/unstable/#function-library-lib.fileset.maybeMissing -->

```{code-block} diff
:caption: build.nix
 { stdenv, lib }:
 let
   fs = lib.fileset;
-  sourceFiles = fs.difference ./. ./result;
+  sourceFiles = fs.difference ./. (fs.maybeMissing ./result);
 in
```

This now works, using the whole directory since `./result` is not present:

```
$ nix-build
trace: /home/user/select (all files in directory)
this derivation will be built:
  /nix/store/zr19bv51085zz005yk7pw4s9sglmafvn-fileset.drv
...
/nix/store/vhyhk6ij39gjapqavz1j1x3zbiy3qc1a-fileset
```

Another build attempt will produce a different trace, but the same output path:

```
$ nix-build
trace: /home/user/fileset
trace: - build.nix (regular)
trace: - default.nix (regular)
trace: - hello.txt (regular)
trace: - nix (all files in directory)
trace: - world.txt (regular)
/nix/store/vhyhk6ij39gjapqavz1j1x3zbiy3qc1a-fileset
```

> ## Union (explicitly exclude files)

Reviewer 3: ```suggestion
## Union for explicit sets
```

Inclusion or exclusion is a matter of how you use it, but union is great for defining sets by explicit enumeration.

Reviewer 3: Oh I get it now. You also have a section for explicit inclusion.
Probably still best to start with that.

Reviewer 2: That would require rewriting it quite a lot, I suggest making that a follow-up, if at all.

Author: Yeah, follow up sounds better!

There is still a problem:
Changing _any_ of the included files causes the derivation to be rebuilt, even though it doesn't depend on those files.

Append an empty line to `build.nix`:

```shell-session
$ echo >> build.nix
```

Again, Nix will start from scratch:

```shell-session
$ nix-build
trace: /home/user/select
trace: - default.nix (regular)
trace: - nix (all files in directory)
trace: - build.nix (regular)
trace: - string.txt (regular)
this derivation will be built:
  /nix/store/zmgpqlpfz2jq0w9rdacsnpx8ni4n77cn-filesets.drv
...
/nix/store/6pffjljjy3c7kla60nljk3fad4q4kkzn-filesets
```

> One way to fix this is to use [`unions`](https://nixos.org/manual/nixpkgs/unstable/#function-library-lib.fileset.unions).

Reviewer 3: ```
One way to fix this is with [`unions`](https://nixos.org/manual/nixpkgs/unstable/#function-library-lib.fileset.unions).
```

but wait, we also need difference?
Maybe start with `unions` without difference? That's actually a great way to make sure you're not including too much. Nix's own flake does that.

Then later flip it around: if you prefer to include any new or previously unknown files by default, you can use `unions` in a `difference` instead.

Author: I found it hard to make the tutorial flow that way, but I can take another look

Author: See https://github.com/NixOS/nix.dev/pull/802#discussion_r1402328626

Create a file set containing a union of the files to exclude (`fs.unions [ ... ]`), and subtract it (`difference`) from the complete directory (`./.`):

```{code-block} diff
:caption: build.nix
 { stdenv, lib }:
 let
   fs = lib.fileset;
-  sourceFiles = fs.difference ./. (fs.maybeMissing ./result);
+  sourceFiles =
+    fs.difference
+      ./.
+      (fs.unions [
+        (fs.maybeMissing ./result)
+        ./default.nix
+        ./build.nix
+        ./nix
+      ]);
 in
```

> Changing any of the excluded files now doesn't necessarily require a rebuild anymore.

Reviewer 3: ```suggestion
Changing any of the excluded files now doesn't necessarily cause a new build anymore.
```

Nix never rebuilds, unless you ask for it.

Reviewer 3: I might have missed another instance of this inaccurate term, because I know I'm also guilty of it.

Check it and modify one of the excluded files again:

```
$ echo >> build.nix
```

```
$ nix-build
trace: /home/user/fileset
trace: - hello.txt (regular)
trace: - world.txt (regular)
/nix/store/ckn40y7hgqphhbhyrq64h9r6rvdh973r-fileset
```

## Filter

> Dealing with a large number of files, independent of their location, can be done programmatically with the [`fileFilter`](https://nixos.org/manual/nixpkgs/unstable/#function-library-lib.fileset.fileFilter) function.

Reviewer 3: ```suggestion
With the [`fileFilter`](https://nixos.org/manual/nixpkgs/unstable/#function-library-lib.fileset.fileFilter) function, you can find sets of files by criteria.
```

Actually close to set builder notation, for the mathematically inclined.

Author: Changing this to "find sets of files satisfying a criteria"

Use it to select all files with a name ending in `.nix`:

```{code-block} diff
:caption: build.nix
 { stdenv, lib }:
 let
   fs = lib.fileset;
   sourceFiles =
     fs.difference
       ./.
       (fs.unions [
         (fs.maybeMissing ./result)
-        ./default.nix
-        ./build.nix
+        (fs.fileFilter (file: lib.hasSuffix ".nix" file.name) ./.)
         ./nix
       ]);
 in
```

> This does not change the result:

Reviewer 3: ```suggestion
This does not change the result, even if we add a new `.nix` file.
```

     fileset = sourceFiles;
   };
   postInstall = ''
-    mkdir $out
-    cp -v {hello,world}.txt $out
+    cp -vr . $out
   '';
 }
```

The `postInstall` script is simplified to rely on the sources to be pre-filtered appropriately:

```shell-session
$ nix-build
trace: /home/user/fileset
trace: - build.sh (regular)
trace: - hello.txt (regular)
trace: - src (all files in directory)
trace: - world.txt (regular)
this derivation will be built:
  /nix/store/sjzkn07d6a4qfp60p6dc64pzvmmdafff-fileset.drv
...
'.' -> '/nix/store/zl4n1g6is4cmsqf02dci5b2h5zd0ia4r-fileset'
'./build.sh' -> '/nix/store/zl4n1g6is4cmsqf02dci5b2h5zd0ia4r-fileset/build.sh'
'./hello.txt' -> '/nix/store/zl4n1g6is4cmsqf02dci5b2h5zd0ia4r-fileset/hello.txt'
'./world.txt' -> '/nix/store/zl4n1g6is4cmsqf02dci5b2h5zd0ia4r-fileset/world.txt'
'./src' -> '/nix/store/zl4n1g6is4cmsqf02dci5b2h5zd0ia4r-fileset/src'
'./src/select.c' -> '/nix/store/zl4n1g6is4cmsqf02dci5b2h5zd0ia4r-fileset/src/select.c'
'./src/select.h' -> '/nix/store/zl4n1g6is4cmsqf02dci5b2h5zd0ia4r-fileset/src/select.h'
...
/nix/store/zl4n1g6is4cmsqf02dci5b2h5zd0ia4r-fileset
```

Only the specified files are used, even when a new one is added:

```shell-session

> $ touch src/select.o

Reviewer 3: ```suggestion
$ touch src/select.o src/README.md
```

Add a README for those who know they already filter by git.

```shell-session
$ nix build path:.
```

However it's also not needed, because by default, `nix build` only allows access to files tracked by Git.
:::

## Intersection

This is where `intersection` comes in.
It allows creating a file set that consists only of files that are in _both_ of the two given file sets.

Select all files that are both tracked by Git *and* relevant for the build:

```{code-block} diff
:caption: build.nix
 { stdenv, lib }:
 let
   fs = lib.fileset;
-  sourceFiles = fs.gitTracked ./.;
+  sourceFiles =
+    fs.intersection
+      (fs.gitTracked ./.)
+      (fs.unions [
+        ./hello.txt
+        ./world.txt
+        ./build.sh
+        ./src
+      ]);
 in
```

This will produce the same output as in the other approach and therefore re-use a previous build result:

```shell-session
$ nix-build
warning: Git tree '/home/user/select' is dirty
trace: - build.sh (regular)
trace: - hello.txt (regular)
trace: - src
trace:   - select.c (regular)
trace:   - select.h (regular)
trace: - world.txt (regular)
/nix/store/zl4n1g6is4cmsqf02dci5b2h5zd0ia4r-fileset
```

## Conclusion

We have shown some examples on how to use all of the fundamental file set functions.

> For more complex behavior, they can be composed as needed.

Reviewer 3: ```suggestion
For more complex use cases, they can be composed as needed.
```

  can address these problems.
  However, it's often hard to express the desired path selection using the `filter` function interface.

In this tutorial you'll learn how to use Nixpkgs' [`lib.fileset` library](https://nixos.org/manual/nixpkgs/unstable/#sec-fileset) to work with local files in derivations.

> It abstracts over built-in functionality with a safer and more convenient interface.

Reviewer 2: I'd revert both sentences. Apostrophising Nixpkgs looks really weird, and `abstracting over X with Y`, while grammatically correct can be misleading to be read as "built in functionality having a safer interface".

## File sets

> The basic concept is that of a _file set_,

Reviewer 2: Revert too (or if you want file set as singular, say "based on the concept of a file set". "The basic concept is" - of what? What's the subject? And starting with "the basic concept" even if you say "of the file set library" is probably not even important, and the original sentence said what we want: that the file set revolves around this basic concept.

Author: In the docs team meeting right now, changing it to

> A _file set_ is a data type representing a collection of local files.
File sets can be created, composed, and manipulated with the various functions of the library.

null
```

> All functions that expect a file set for an argument also accept a [path](https://nixos.org/manual/nix/stable/language/values#type-path) instead.

Reviewer 2: Revert, because "instead" can be misinterpreted as "mutually exclusive".

Author: Changing it to

> All functions that expect a file set for an argument can also accept a [path](https://nixos.org/manual/nix/stable/language/values#type-path).

:::

Even though file sets conceptually contain local files, these files are *never* added to the Nix store unless explicitly requested.

> So You don't have to worry as much about accidentally copying secrets into the world-readable store.

Reviewer 2: Capitalisation. Not sure if the connective is really needed here. I would (and did) skip it.

Author: Changing to

> Because of that you don't have to worry as much about accidentally copying secrets into the world-readable store.

Since `src` refers to the whole directory, and its contents change when `nix-build` succeeds, Nix will have to start over every time.

:::{note}

> This will also happen with `src = ./.`, without using the file set library.

Reviewer 2: This sentence structure is potentially confusing. What you probably mean here is "... i.e., without using the file set library", but that makes it a lot more clunky. I'd revert that too.
